File: Programmers/[136797] 숫자 타자 대회/ONE/136797.java
import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

class Solution {
    static class Node {
        private final int x;
        private final int y;
        private final int weight;

        public Node(int x, int y, int weight) {
            this.x = x;
            this.y = y;
            this.weight = weight;
        }

        public int x() {
            return x;
        }

        public int y() {
            return y;
        }

        public int weight() {
            return weight;
        }
    }

    private final int[] dx = {-1, -1, -1, 0, 1, 1, 1, 0};
    private final int[] dy = {-1, 0, 1, 1, 1, 0, -1, -1};
    private final int[][] keys = new int[][]{{1, 2, 3}, {4, 5, 6}, {7, 8, 9}, {-1, 0, -1}};
    private final int[][] weights = new int[10][10];
    private int[][][] dp;

    public int solution(String numbers) {

        for (int[] w : weights) {
            Arrays.fill(w, -1);
        }

        for (int row = 0; row < 4; row++) {
            for (int col = 0; col < 3; col++) {
                if (keys[row][col] == -1) {
                    continue;
                }
                setWeight(row, col);
            }
        }
        for (int index = 0; index < 10; index++) {
            weights[index][index] = 1;
        }

        int length = numbers.length();
        dp = new int[length + 1][10][10];
        for (int number = 0; number < length + 1; number++) {
            for (int left = 0; left < 10; left++) {
                Arrays.fill(dp[number][left], -1);
            }
        }
        return simulate(0, length, 4, 6, numbers);
    }

다

    private boolean isInbound(int x, int y) {
        return x >= 0 && y >= 0 && x < 4 && y < 3;
    }

    private int simulate(int index, int length, int left, int right, String numbers) {
        if (index == length) {
            return 0;
        }
        if (dp[index][left][right] != -1) {
            return dp[index][left][right];
        }

        int number = Character.getNumericValue(numbers.charAt(index));
        int result = Integer.MAX_VALUE;

        if (number != right) {
            result = Math.min(result,
                    simulate(index + 1, length, number, right, numbers) + weights[left][number]);
        }
        if (number != left) {
            result = Math.min(result,
                    simulate(index + 1, length, left, number, numbers) + weights[right][number]);
        }
        return dp[index][left][right] = result;
    }
}

File: Programmers/[136797] 숫자 타자 대회/KingDonggyu/136797.js
function calculateTime(hand, target) {
  // 제자리에서 누르기
  if (hand === target) {
    return 1;
  }

  const keyboard = {
    1: [0, 0],
    2: [0, 1],
    3: [0, 2],
    4: [1, 0],
    5: [1, 1],
    6: [1, 2],
    7: [2, 0],
    8: [2, 1],
    9: [2, 2],
    0: [3, 1],
  };

  let x = Math.abs(keyboard[hand][0] - keyboard[target][0]);
  let y = Math.abs(keyboard[hand][1] - keyboard[target][1]);

  let time = 0;

  // 대각선으로 이동하기
  while (x && y) {
    x -= 1;
    y -= 1;
    time += 3;
  }

  // 상하좌우로 이동하기
  time += (x + y) * 2;

  return time;
}

function getTimes() {
  const result = [];

  for (let hand = 0; hand < 10; hand++) {
    const times = [];

    for (let target = 0; target < 10; target++) {
      times.push(calculateTime(hand, target));
    }

    result.push(times);
  }

  return result;
}

function solution(numbers) {
  const times = getTimes();
  const dp = Array.from(Array(numbers.length + 1), () =>
    Array.from(Array(10), () => Array(10).fill(Infinity))
  );

  dp[0][4][6] = 0;

  for (let i = 0; i < numbers.length; i++) {
    const num = numbers[i];

    for (let left = 0; left < 10; left++) {
      for (let right = 0; right < 10; right++) {
        const prevTime = dp[i][left][right];

        if (left === right || prevTime === Infinity) {
          continue;
        }

        if (dp[i + 1][num][right] > prevTime + times[left][num]) {
          dp[i + 1][num][right] = prevTime + times[left][num];
        }

        if (dp[i + 1][left][num] > prevTime + times[right][num]) {
          dp[i + 1][left][num] = prevTime + times[right][num];
        }
      }
    }
  }

  return Math.min(...dp[numbers.length].flat());
}
